File: csharp/LeetCode.Questions/src/Exercises/Solutions/BinarySearch.cs
namespace Exercises.Solutions;

    public class BinarySearch
    {
        public static int search(int[] nums, int target)
        {
            int low = 0; // Inicializa o índice mais baixo do array
            int high = nums.Length - 1; // Inicializa o índice mais alto do array

            // Realiza a busca enquanto o índice mais baixo não ultrapassa o mais alto
            while (low <= high)
            {
                var mid = (low + high) / 2; // Calcula o índice do elemento do meio

                if (nums[mid] == target) // Se o elemento do meio for o alvo, retorna o índice
                    return mid;
                else if (nums[low] <= nums[mid]) // Verifica se a metade esquerda está ordenada
                {
                    if (target > nums[mid] || target < nums[low]) // Verifica se o alvo está na metade direita
                    {
                        low = mid + 1; // Descarta a metade esquerda, incluindo o meio
                    }
                    else
                    {
                        high = mid - 1; // Descarta a metade direita, mantendo o elemento do meio
                    }
                }
                else // Caso contrário, a metade direita está ordenada
                {
                    if (target < nums[mid] || target > nums[high]) // Verifica se o alvo está na metade esquerda
                    {
                        high = mid - 1; // Descarta a metade direita, mantendo o elemento do meio
                    }
                    else
                    {
                        low = mid + 1; // Descarta a metade esquerda, incluindo o meio
                    }
                }
            }
            return -1; // Retorna -1 se o alvo não for encontrado no array
        }
    }

File: go/exercises/code/BinarySearch.go
package code

// Search realiza uma busca binária modificada em um array rotacionado
// Retorna o índice do target se encontrado, ou -1 se o target não estiver presente
func Search(nums []int, target int) int {

	low := 0                 // Inicializa o limite inferior
	high := len(nums) - 1    // Inicializa o limite superior

	// Continua enquanto a faixa de busca for válida
	for low <= high {
		// Calcula o índice do meio
		mid := (low + high) / 2

		// Verifica se o elemento do meio é o alvo
		if nums[mid] == target {
			return mid // Retorna o índice se encontrado
		}

		// Verifica se a metade esquerda do array está ordenada
		if nums[low] <= nums[mid] {
			// Verifica se o alvo está dentro da metade esquerda ordenada
			if target > nums[mid] || target < nums[low] {
				// Move o limite inferior para a direita do meio
				low = mid + 1
			} else {
				// Ajusta o limite superior para procurar na metade esquerda
				high = mid - 1
			}
		} else {
			// A metade direita do array está ordenada
			if target < nums[mid] || target > nums[high] {
				// Ajusta o limite superior para procurar na metade esquerda
				high = mid - 1
			} else {
				// Move o limite inferior para a direita para procurar na metade direita
				low = mid + 1
			}
		}
	}

	// Se o alvo não foi encontrado, retorna -1
	return -1
}
